### DISCUSSION TITLE
[Documentation] (1) jruby versus ruby one click installer, and (2) GraalVM

### DISCUSSION THREAD

**Discussion by rubyFeedback** [jruby/jruby#6804] [Q&A] [answered]
Hey guys,

Sorry to do two issues in one but I felt it was easier. The first issue is more important;
second one is more interesting in the long term perhaps.

Anyway.

(1) I use Linux most of the time but I have a win10 laptop. I recently ran into a
problem with my own code. That's fine - my fault. I was requiring a gem that 
was not properly loaded, and then assumed the namespace was available.
So I fixed that code. But then I looked at the error message and noticed that
it came from jruby, and it was still referring the old code.

Again - user error here. I forgot that I had jruby, and that this jruby was installed
and in use. I mistakingly assumed it was the one click installer (ruby 3.x something).

Of course we can check via RUBY_VERSION. So, this is ultimately all coming
down to a user error.

But!

Perhaps I am not the only one who may have had more than one ruby version
installed? My code works fine on ruby 2.5.8 too, and if not I can adjust - no
problem. But I simply mistakingly assumed I was updating my gems when
instead I was using another ruby version (jruby).

So this brings me to my question and the main reason for this issue - I hope
the above can serve as the use case explanation for the issue.

- Does the jruby team recommend certain setups? E. g. jruby only? Is that
realistic?

I would very much appreciate if someone who has knowledge about the jruby
situation and ecosystem could explain this a little bit.

I batch-install my own gems anyway and I have a .rb file that does this,
so the fix for my situation was simply to use "jruby" rather than "ruby"
(I still have the one click installer available under c:\ruby). But I am wondering
whether it would not be simpler for my case to uninstall one or the other
version - just to keep it simple. Perhaps others have a better use case.

This is what I would love to read in the documentation.

https://www.jruby.org/documentation is quite sparse, so perhaps in the
long run someone could explain their own setup/use cases, so that
downstream users could learn. I assume most folks use linux or 
mac, but some may (have to) use windows, and in this case it could
be useful to know the workflow.

Which brings me to ...

(2) GraalVM

Eregon probably knows a lot about it due to truffle ruby and what not.
GraalVM seems to have a lot of momentum, but I wonder how all these
ruby versions all play together. How many ruby versions do we have
to use? :)

At the least to me this is all a bit complicated right now. Yes, a lot of
this is due to me simply not knowing right now... I used Linux for too
long! And there it's super-simple for me: just put different ruby
versions under /home/Programs/Ruby/ - the GoboLinux approach. :P

On windows this is a little bit more complicated and the more ruby
versions the more confusing. So if the jruby team could add to the
documentation in regards to GraalVM this would be nice too. Does
GraalVM obsolete jruby in the long run? Questions like this.

Sorry to be docu-annoying but I really feel jruby could benefit from
a better documentation (it may be all available elsehwere or on
the wiki, but I go to jruby.org so simply having pointers and links
may be helpful. It's also ok if these become outdated at a later
time; I read old blogs and adjust the information therein accordingly,
but even old information is often better than no information at all
whatsoever).

Don't worry about (2) too much though - (1) is really the main gist
of this issue request. Thank you for reading and potentially
considering! Please feel free to close this issue at any moment in
time.

**Comment by enebo** [comment, marked as answer]
@rubyFeedback I agree that we should improve our documentation.  If you feel so inclined perhaps after some comments then you can maybe adjust our wiki a little bit?  We do make efforts but we are fairly burdened with just updating JRuby itself.  It is a time-old problem that documentation gets neglected.

The windows issue and having multiple versions is the same issue if you had two versions of C Ruby installed.  I have never been asked about maintaining both before for windows itself but it is too bad projects like https://github.com/vertiginous/pik did not succeed (although perhaps it still works ok).  This would make working with multiple Ruby versions much simpler.

Here is one thought: remove bin/ruby.bat from JRuby install so you when you type ruby you will get one-click and when you type jruby you will obviously run JRuby.

Another thing is you can guard against accidentally running a particular script by checking RUBY_ENGINE.  It is more what you want than RUBY_VERSION.

The other somewhat obvious solution would be to make a bat script to switch between Rubies by updating PATH (which likely will contain something like %RUBY_LOCATION% in the %PATH% variable.  I think with power shell you could even print out which ruby you were using using env vars in your prompt?  I am not a windows user tbh.  I half wonder if this does exist for some Ruby user who has n Ruby versions they want to test windows with.

If you figure out or someone else has better recommendations we can create a wiki page on this.

As for things like how do we compare to competing projects like TruffleRuby or CRuby I think we definitely should make that and probably engage those projects to make sure we are not saying something they will object to.  GraalVM itself can run JRuby fine as a JVM but to answer the only question you posed (as an example) no TruffleRuby will not obsolete JRuby.  We do not have a complete set of overlapping features (like we have a goal to run well on all JVMs) and we have a lot of production users.  I agree though we definitely need to address this confusion in particular since we get asked this quite a bit.

**Reply by headius** [reply]
I marked this as the "answer" because it covers both aspects of the original post, but @mohits's post below also illustrates how to have both JRuby and CRuby install.

**Comment by mohits** [comment]
@rubyFeedback - thanks for the question.

> Perhaps I am not the only one who may have had more than one ruby version
installed? My code works fine on ruby 2.5.8 too, and if not I can adjust - no
problem. But I simply mistakingly assumed I was updating my gems when
instead I was using another ruby version (jruby).

I work mainly on Windows 10 and want to help people who are facing problems like yours on Windows :-)
I strongly recommend pik for multiple Ruby installations.
[1] Using pik - https://notepad.onghu.com/2021/ruby3-windows-pik-multiple-installations/
[2] Installing and using Ruby, including setup for pik
-> [JRuby] https://notepad.onghu.com/2020/jruby-win-day0-install-hello_world/ 
-> [CRuby] https://notepad.onghu.com/2021/ruby3-windows-day0-install-hello_world/

**Comment by enebo** [comment]
@mohits Thanks for that post recommending pik.  I made the mistake of assuming a project with no commits for years was dead (or unworking).

**Comment by headius** [comment]
Since there's no real issue here I have converted this to a discussion. Thank you for raising these questions!
